### PR TITLE
Implemented widget drag & drop for side panels

### PR DESCRIPTION
**Pull request by spoenemann**
* [x] Drag widgets into expanded side panels
* [x] Drag widgets out of side panels
* [x] Drag widgets into collapsed side panels
* [x] Collapse side panels by clicking on a tab
* [x] Reorder tabs in side bars

### CODE REVIEW

**Comment by spoenemann** [comment]
* [ ] Replace previous bottom panel implementation with a multi-document dock panel (same as in main area)
 * [ ] Fix #1126

**Comment by svenefftinge** [comment]
Is it ready for review? Could please rebase, squash (where it makes sense) and sign-off the commits?

**Comment by spoenemann** [comment]
Ready.

File: packages/core/src/browser/shell/side-panel-handler.ts [hunk 1]
```diff
@@ -255,24 +470,158 @@ export namespace SidePanel {
...
+ * side panels of the application shell.
+ */
+export class TheiaDockPanel extends DockPanel {
+
+    private __drag?: Drag;
+    private attachDone = false;
+
+    constructor(options?: DockPanel.IOptions) {
+        super(options);
+        // Override the _drag property from DockPanel with an accessor property
```
Comment (spoenemann) [inline packages/core/src/browser/shell/side-panel-handler.ts]: This dirty hack could be eliminated if phosphorjs/phosphor#326 is merged.

File: packages/core/src/browser/shell/tab-bars.ts [hunk 2]
```diff
@@ -66,3 +165,208 @@ export class TabBarRenderer implements TabBar.IRenderer<any> {
...
+    /**
+     * Tab bars of the left and right side panel are arranged vertically by rotating their labels.
+     * Rotation is realized with the CSS `transform` property, which disrupts the browser's ability
+     * to arrange the involved elements automatically. Therefore the elements are arranged explicitly
+     * by the TabBarRenderer using inline `height` and `top` styles. However, the size of labels
+     * must still be computed by the browser, so the rendering is performed in two steps: first the
+     * tab bar is rendered horizontally inside a _hidden content node_, then it is rendered again
+     * vertically inside the proper content node. After the first step, size information is gathered
+     * from all labels so it can be applied during the second step.
+     */
```
Comment (spoenemann) [inline packages/core/src/browser/shell/tab-bars.ts]: For more information on the problems caused by the CSS `transform` attribute see https://drafts.csswg.org/css-transforms/#transform-rendering :

> Transformations do affect the visual rendering, but have no affect on the CSS layout other than affecting overflow.

**Comment by svenefftinge** [comment]
We should add an icon to the tabbar of the bottom panel, to trigger `Toggle Bottom Panel`.
Otherwise awesome work! :)

**Comment by spoenemann** [comment]
> We should add an icon to the tabbar of the bottom panel, to trigger `Toggle Bottom Panel`.

See #1168

**Comment by hexa00** [comment]
Just tried this out, it's awesome! nice work!

The only thing I found a bit confusing is dragging the windows to a side collapsed area...  I expected to drag to the collapsed sidebar but it expands let's say Git.. and I drop it there to add it to the sidebar... that's a bit weird to me but nothing major.

**Comment by hexa00** [comment]
Just found another weird thing... since the terminal is passing through keys like Ctrl+J if you have the terminal in focus you can't toggle the bottom panel off...

Not sure how to fix this :(

**Comment by svenefftinge** [comment]
@hexa00 toggling the bottom panel is discussed here https://github.com/theia-ide/theia/issues/1168

**Comment by spoenemann** [comment]
> The only thing I found a bit confusing is dragging the windows to a side collapsed area... I expected to drag to the collapsed sidebar but it expands let's say Git.. and I drop it there to add it to the sidebar... that's a bit weird to me but nothing major.

I had an intermediate version where it was possible to drop into the side bar. That felt a bit alien because the DockPanel that is used in the main area (and now also in the bottom area) don't allow that. I rolled back that approach and now the dropping is done entirely with the existing implementation of DockPanel. I would say that this way we have a more consistent behavior than if we would allow dropping into tab bars here but don't allow it there.

**Comment by hexa00** [comment]
@spoenemann OK I agree it's better to be consistent.

**Comment by akosyakov** [comment]
@spoenemann please resolve conflicts

**Comment by akosyakov** [comment]
@hexa00 can you review and approve it is fine? code changes are fine with me

**Comment by svenefftinge** [comment]
This is very good and should be merged. @spoenemann Could you rebase?

**Comment by spoenemann** [comment]
> Could you rebase?

&#9989;

**Pull request merged**